File: app/app/src/main/java/com/example/bankmanagement/base/fragment/BaseFragment.java
package com.example.bankmanagement.base.fragment;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.LifecycleObserver;

import com.example.bankmanagement.base.BaseActions;
import com.example.bankmanagement.base.activity.BaseFragmentActivity;
import com.example.bankmanagement.base.navigation.FragmentNavigator;

public abstract class BaseFragment
        extends Fragment implements BaseActions, LifecycleObserver {

    protected View rootView;
    private Context fragmentContext;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        fragmentContext = context;
        getLifecycle().addObserver(this);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        fragmentContext = null;
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        rootView = LayoutInflater.from(fragmentContext)
                .inflate(setLayout(), container, false);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        initValues();
        initViews();
        initActions();
        fetchData();
    }

    protected void initViews() {

    }

    protected void initBackButtons(View... views) {
        for (View view : views) {
            onBackPress();
        }
    }

    protected abstract void fetchData();

    public Context getFragmentContext() {
        return fragmentContext;
    }


    /**
     * Get fragmentNavigator
     */
    public FragmentNavigator getNavigator() {
        return ((BaseFragmentActivity) fragmentContext).getNavigator();
    }

    /**
     * Get active fragment
     *
     * @return
     */
    protected Fragment getActiveFragment() {
        return getNavigator().getActiveFragment();
    }

    /**
     * Set fragment to a Container view
     *
     * @param containerViewId
     * @param fragment
     */
    protected void setFragmentToContainer(@IdRes int containerViewId, @NonNull Fragment fragment) {
        FragmentTransaction transaction = getChildFragmentManager().beginTransaction();
        transaction.add(containerViewId, fragment).commit();
    }

    /**
     * Check if fragment is already in ContainerView then show fragment
     * else add new one to
     *
     * @param containerViewId
     * @param fragment
     */
    protected void showOrAddFragment(
            @IdRes int containerViewId,
            Fragment fragment
    ) {
        showOrAddFragment(null, containerViewId, fragment);
    }

    /**
     * Check if fragment is already in ContainerView then show fragment
     * else add new one to
     *
     * @param containerViewId
     * @param fragment
     */
    protected void showOrAddFragment(
            FragmentManager fragmentManager,
            @IdRes int containerViewId,
            Fragment fragment
    ) {
        if (fragment == null) return;

        if (fragmentManager == null)
            fragmentManager = getChildFragmentManager();

        String fragmentTag = getFragmentTag(fragment);

        Fragment fragmentByTag = fragmentManager.findFragmentByTag(fragmentTag);
        if (fragmentByTag != null) {
            //if the fragment exists, show it.
            fragmentManager.beginTransaction().show(fragmentByTag).commit();
        } else {
            //if the fragment does not exist, add it to fragment manager.
            fragmentManager.beginTransaction().add(containerViewId, fragment, fragmentTag).commit();
        }
    }

    protected void hideFragment(
            Fragment fragment
    ) {
        if (fragment == null) return;

        FragmentManager fragmentManager = getChildFragmentManager();
        String fragmentTag = getFragmentTag(fragment);

        if (fragmentManager.findFragmentByTag(fragmentTag) != null) {
            //if the other fragment is visible, hide it.
            fragmentManager.beginTransaction().hide(fragment).commit();
        }
    }

    @NonNull
    private String getFragmentTag(@NonNull Fragment fragment) {
        return fragment.getClass().getSimpleName();
    }

    /**
     * The extend fragment can overRide this to handle the Activity backPress
     *
     * @return true --> swallow the back press
     * false --> keep it
     */
    public boolean onBackPress() {
        return false;
    }
}

File: app/app/src/main/java/com/example/bankmanagement/base/fragment/BaseFragment.kt
package com.hanheldpos.ui.base.fragment

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.observe
import com.example.bankmanagement.base.activity.BaseActivity
import com.example.bankmanagement.base.dialog.AppAlertDialog
import com.example.bankmanagement.base.viewmodel.BaseViewModel
import com.example.bankmanagement.base.viewmodel.ViewState

abstract class BaseFragment<T : ViewDataBinding, VM : BaseViewModel> : Fragment() {

    protected lateinit var fragmentContext: Context

    // Binding
    protected lateinit var binding: T

    @LayoutRes
    protected abstract fun layoutRes(): Int

    // ViewModel
    protected abstract val viewModel:VM
    protected abstract fun viewModelClass(): Class<VM>
    protected abstract fun initViewModel(viewModel: VM)

    // Data & Actions
    protected abstract fun initView()
    protected abstract fun initData()
    protected abstract fun initAction()

    override fun onAttach(context: Context) {
        super.onAttach(context)
        fragmentContext = context
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(
            LayoutInflater.from(fragmentContext),
            layoutRes(),
            container,
            false
        )
        binding.apply {
            root.setOnTouchListener { v: View, _: MotionEvent? ->
                v.isClickable = true
                v.isFocusable = true
                false
            }
        }


        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // ViewModel
        viewModel
            .apply {
                initViewModel(this)
            }

        initView()
        initData()
        initAction()

        viewModel.viewState.observe(viewLifecycleOwner, { viewState ->
            viewState?.run {
                when (viewState) {
                    ViewState.SHOW_LOADING -> showLoading(true)
                    ViewState.HIDE_LOADING -> showLoading(false)
                    ViewState.SHOW_ERROR -> viewModel.errMessage?.run {
                        showMessage(this)
                    }
                }
            }
        })
    }




    /**
     * Show loading dialog
     */
    fun showLoading(show: Boolean) {
        (activity as? BaseActivity<*, *>)?.showLoading(show)
    }

    /**
     * Show error dialog
     */
    private fun showError() {
        (activity as? BaseActivity<*, *>)?.showError()
    }

    /**
     * Show error dialog
     */
    fun showMessage(message: String?) {
        showAlert(message = message)
    }

    /**
     * Show alert dialog with full options
     * ------------------------------------------
     * | Title                                  |
     * | Message                                |
     * ------NEGATIVE BUTTON---POSITIVE BUTTON---
     */
    fun showAlert(
        title: String? = null,
        message: String?,
        positiveText: String? = null,
        negativeText: String? = null,
        onClickListener: AppAlertDialog.AlertDialogOnClickListener? = null
    ) {
        (activity as? BaseActivity<*, *>)?.showAlert(
            title,
            message,
            positiveText,
            negativeText,
            onClickListener
        )
    }

    /**
     * Request permission
     */
    fun requestPermission(
        explainPermission: String,
        permissions: List<String>,
        callback: BaseActivity.RequestPermissionCallback
    ) {
        (activity as? BaseActivity<*, *>)?.requestPermission(
            explainPermission,
            permissions,
            callback
        )
    }


}